feat: RP for plugin-tee-verifiable-log-api

gene-zhan opened the pull request:
# Relates to:

1. Run verifiable Eliza in the TEE.
2. Verifiable log for Eliza in the TEE
3.  https://github.com/ai16z/eliza/pull/1259

# Risks

Low

# Background

## What does this PR do?

This PR builds upon **`plugin-tee-verifiable-log`** by modifying the direct client to add remote attestation and query interfaces for verifiable logs.

To better understand what verifiable logs are and why we implemented this feature, it is necessary to refer to the preceding PR that introduces **`plugin-tee-verifiable-log`** and provides the context for its development.

## What kind of change is this?

- Feature
- TEE plugin
- Verifiable feature

# Documentation changes needed?

Yes, we will add documentation about api detail.

# Testing

## Where should a reviewer start?

1. **Understand the existing `plugin-tee`**
    1. We rely on Eliza's existing `plugin-tee` and use its key derivation interface.
    
2. To better understand what verifiable logs are and why we implemented this feature, it is necessary to refer to the preceding PR that introduces **`plugin-tee-verifiable-log`** and provides the context for its development.
    
    
3. **Understand what `plugin-tee-verifiable-log` does**
    1. **Derive a key pair for verifiable logs**: It uses the TEE to derive a key pair specifically for signing verifiable logs.
    2. **Remote attestation**: The public key of verifiable logs is embedded in the remote attestation report, making it accessible for external verification.
    3. **Sign logs**: This plugin accepts logs passed in from external modules (e.g., tweets sent and received by a Twitter client, or executed actions), signs them with the TEE-derived key pair, and stores them in the database.
    4. **Verification**: External entities can use the remotely attested public key to verify these logs, ensuring that certain actions were indeed performed by TEE Eliza.
    

## Detailed testing steps

It have completed the integration tests and can run the `pnpm test` file in the test directory.

Commit:
feat: document advanced usage in md file

Commit:
feat: after CR, removing long example

mcp97 committed:
Merge pull request #1136 from erise133/feat/handlebars-templating-engine

feat: add support for handlebars templating engine as an option

jzvikart committed:
Get stopAgent() to work as expected

jzvikart committed:
Cleanup

jzvikart committed:
Avoid using console.clear() to preserve messages for debugging

rarepepi committed:
remove docker compose command since Docker file already runs

azep-ninja committed:
fix tg duplicate functions.

owlcode committed:
clean newlines for new tweet

mcp97 committed:
Merge pull request #1141 from owlcode/clean-newlines-from-new-tweets

clean newlines for new tweet

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

odilitime committed:
Merge branch 'develop' into fix/plugin-evm

mcp97 committed:
Merge pull request #1140 from azep-ninja/fix/duplicate-tg-funtions

fix: telegram client duplicate function removal

odilitime committed:
fix merge: remove double improve, adjust params to various calls, use initWalletProvider

odilitime committed:
Merge pull request #965 from FWangZil/fix/plugin-evm

fix: Fix Parameter Parsing in plugin-evm TransferAction and Return Transaction Hash

odilitime committed:
Merge pull request #1135 from fomoTon/fomo-token-plugin

feat: allow agents to create/buy/sell tokens on FOMO.fund's bonding curve in plugin-solana

odilitime committed:
update lockfile for PR1135

shakkernerd committed:
chore: remove comment

shakkernerd committed:
Merge pull request #1143 from ai16z/remove-comment

chore: remove comment

shakkernerd committed:
fix: remove unused import

shakkernerd committed:
fix: Expected an assignment or function call and instead saw an expression

shakkernerd committed:
fix: use of let instead of const

shakkernerd committed:
chore: comment out unused imports

shakkernerd committed:
fix: remove unused variable

shakkernerd committed:
chore: lint command

shakkernerd committed:
chore: add eslint config

shakkernerd committed:
chore: delete client-whatsapp folder

mcp97 committed:
Merge pull request #1139 from rarepepi/docker-fixes

fix: remove docker compose command since Docker file already runs

odilitime committed:
improve on fomo plugin and distingush it from pump.fun's plugin

shakkernerd committed:
Merge pull request #1147 from odilitime/fix-lint

fix: improve fomo integration

odilitime committed:
include fomo

odilitime committed:
update URLs and example response

odilitime committed:
Merge branch 'develop' of https://github.com/ai16z/eliza into fix-lint

shakkernerd committed:
Merge pull request #1148 from odilitime/fix-lint

chore: fix PR #1147

odilitime committed:
Merge pull request #1144 from ai16z/develop

chore: Merge monday, merging develop into main

actions-user committed:
chore: update changelog

shakkernerd committed:
feat: update packages version script

jzvikart committed:
Merge remote-tracking branch 'ai16z/develop' into realitySpiral/integration-tests

shakkernerd committed:
Merge pull request #1150 from ai16z/update-version

feat: update packages version script

jzvikart committed:
Remove dependency zx

shakkernerd committed:
chore: bump version to 0.1.6-alpha.3

shakkernerd committed:
Merge pull request #1152 from ai16z/new_version

chore: bump version to 0.1.6-alpha.3

jzvikart committed:
Minor cleanup

shakkernerd committed:
fix: fetch log level to debug

shakkernerd committed:
Merge pull request #1153 from ai16z/fetch_logs_debug

fix: fetch log level to debug

odilitime committed:
fix directClient

mcp97 committed:
Merge pull request #1154 from odilitime/fix-lint

fix: fix direct-client ability to start agents

shakkernerd committed:
Merge pull request #1155 from ai16z/develop

chore: develop into main

actions-user committed:
chore: update changelog

jzvikart committed:
Disable logging of fetch requests to avoid disclosing sensitive information

madjin committed:
update community docs

tcm390 committed:
allow multiple bots to join the voice channel

tcm390 committed:
allow multiple bots to join the voice channel

shakkernerd committed:
chore: fix broken pnpm lockfile

shakkernerd committed:
fix: client twitter login and auth handler

shakkernerd committed:
Merge pull request #1158 from ai16z/fix/client-twitter

fix: client twitter login and auth handler

yang-han committed:
fix: print commands to start the client and remove unused --non-iteractive in dockerfile

yang-han committed:
fix: print commands to start the client and remove unused --non-iteractive in dockerfile

shakkernerd committed:
chore: bump version to 0.1.6-alpha.4

shakkernerd committed:
Merge pull request #1159 from ai16z/new_version

chore: bump version to 0.1.6-alpha.4

actions-user committed:
chore: update changelog

shakkernerd committed:
Merge pull request #1156 from ai16z/tcm-discord-multiple-bot

fix: Enable multiple bots to join Discord voice channels

mcp97 committed:
Merge pull request #1163 from yang-han/fix_client

chore: print commands to start the client and remove unused --non-itera…

shakkernerd committed:
feat: make script dash compatible

shakkernerd committed:
Merge pull request #1165 from ai16z/fix/start_script

feat: make script dash compatible

thomasWos committed:
Fix typo

YoungPhlo committed:
docs: Update "What Did You Get Done This Week? 5" spaces notes

jzvikart committed:
Change CI trigger to 'pull_request'

odilitime committed:
Merge branch 'develop' into main

odilitime committed:
Merge pull request #1170 from thomasWos/main

fix: Fix typo in multiversx plugin prompt for creating token

odilitime committed:
Merge branch 'develop' into docs/community-streams

odilitime committed:
Merge pull request #1174 from YoungPhlo/docs/community-streams

docs: Update "What Did You Get Done This Week? 5" spaces notes

SumeetChougule committed:
Fix client.push issue and update README for Slack client verification

ai16z-demirix committed:
test: adding tests for runtime.ts. Modified README since we switched to vitest

Commit:
fixed CONTRIBUTING.md file Issue: 1048

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

mcp97 committed:
Merge pull request #1191 from ileana-pr/main

docs: fixed CONTRIBUTING.md file Issue: 1048

mcp97 committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1190 from ai16z-demirix/main

test: adding tests for runtime.ts. Modified README since we switched to vitest

gene-zhan committed:
Merge branch 'ai16z:main' into main

tomguluson92 committed:
Update README_CN.md

Add more details in the CN version of README

9547 committed:
docs(cn): add python2.7

Signed-off-by: 9547 <29431502+9547@users.noreply.github.com>

9547 committed:
docs(cn): rm duplicated -

Signed-off-by: 9547 <29431502+9547@users.noreply.github.com>

jzvikart committed:
Merge remote-tracking branch 'ai16z/develop' into jure/integration-tests

netdragonx committed:
Fix: optional chaining on search to prevent error when search is uncommented but not enabled in env file.

tobbelobb committed:
Write summary file before trying to cache it

 - Also give a .md file extension for prettier rendering in Discord

v1xingyue committed:
gitpod cicd bug

Sometimes we can't fetch tags, so let's fetch tags before.

shakkernerd committed:
Merge pull request #1177 from Sifchain/jure/integration-tests

feat: integration tests fixes + library improvements

marcNY committed:
Update README.md

Fix starter script: Add missing ‘cd’ command to navigate to ‘eliza-starter’ directory

Commit:
Delete scripts/gettweets.mjs

odilitime committed:
Merge pull request #1201 from 9547/doc-cn

docs(cn): add python 3.7

odilitime committed:
Merge branch 'develop' into main

odilitime committed:
Merge pull request #1207 from v1xingyue/v1xingyue-patch-1

fix: gitpod cicd bug

odilitime committed:
Merge branch 'develop' into patch-1

odilitime committed:
Merge pull request #1209 from marcNY/patch-1

docs: Update README.md

odilitime committed:
add missing change directories

odilitime committed:
Merge branch 'develop' into patch-1

odilitime committed:
Merge pull request #1196 from tomguluson92/patch-1

docs: Update "CN README" with more details

SumeetChougule committed:
fixed gitignore

madjin committed:
add wip docs

SumeetChougule committed:
Merge branch 'develop'

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

madjin committed:
update api docs

mcp97 committed:
Merge pull request #1211 from ai16z/new_docs

vpavlin committed:
fix: fail when cannot get token, add Akash to generateText switch

cygaar committed:
Improve generation prompt

Commit:
Rename gettweets-copy.mjs to gettweets.mjs

Commit:
Delete scripts/tweetextractor.mjs

cygaar committed:
Update

odilitime committed:
postgres needs the user to exist before you can add a participant

cygaar committed:
Merge pull request #1217 from cygaar/improve_twitter

fix: improve twitter post generation prompt

tcm390 committed:
support image message

tcm390 committed:
check content type

tcm390 committed:
correct content type

odilitime committed:
Merge branch 'main' into postgres-fix

tcm390 committed:
clean code

tcm390 committed:
handle http image

mcp97 committed:
Merge pull request #1220 from ai16z/tcm-telegram-image

fix: Allow the bot to post messages with images generated by the imageGenerationPlugin on Telegram.

mcp97 committed:
Merge pull request #1219 from odilitime/postgres-fix

fix: postgres needs the user to exist before you can add a participant

tomguluson92 committed:
feat: CircuitBreaker.ts

simpletrontdip committed:
feat: add parse mode, support telebot output in markdown

shakkernerd committed:
Merge branch 'develop' into patch-2

shakkernerd committed:
Merge pull request #1226 from tomguluson92/patch-2

fix: CircuitBreaker.ts

shakkernerd committed:
Merge branch 'develop' into twitter-credentials

shakkernerd committed:
Merge pull request #1218 from danbednarski/twitter-credentials

chore: clean up scripts

shakkernerd committed:
feat: include akash to handleProvider

shakkernerd committed:
Merge branch 'develop' into fix/akash

simpletrontdip committed:
Merge branch 'develop' into enhance-tele-client

shakkernerd committed:
Merge pull request #1214 from vpavlin/fix/akash

fix: fail when cannot get token, add Akash to generateText switch

shakkernerd committed:
Merge pull request #1229 from simpletrontdip/enhance-tele-client

feat: add parse mode=Markdown, enhance telegram bot output

azurwastaken committed:
fix type in french Readme

azurwastaken committed:
Update README_ES.md

azurwastaken committed:
Update README_IT.md

azurwastaken committed:
Merge branch 'develop' into fix/typo

yjshi2015 committed:
Enhance terminology in the project for clarity and inclusivity

jonathangus committed:
Fix: Sync UI Client with server port env

GeneFuture committed:
add tee-verifiable-log plugin

GeneFuture committed:
add tee-verifiable-log plugin

luisalrp committed:
Fix visibility issue github image cicd

tcm390 committed:
treat twitterShouldRespondTemplate as string

GeneFuture committed:
add tee-verifiable-log plugin api

renlulu committed:
feat: make twitter retry limit a env

renlulu committed:
chore: env update

renlulu committed:
feat: make express payload limit configurable

renlulu committed:
Merge branch 'develop' into feat/env-twitter-retry

renlulu committed:
Merge branch 'develop' into feat/express-payload-config

tomguluson92 committed:
Update generation.ts

tomguluson92 committed:
Update index.ts

tomguluson92 committed:
Update .env.example

odilitime committed:
handle llama_local getTokenForProvider properly

mcp97 committed:
Merge pull request #1244 from renlulu/feat/env-twitter-retry

feat: make twitter login retry times as env

mcp97 committed:
Merge branch 'develop' into develop

mcp97 committed:
Merge pull request #1239 from jonathangus/develop

fix: Sync UI Client with server port env

mcp97 committed:
Merge branch 'develop' into fix/typo

mcp97 committed:
Merge pull request #1236 from azurwastaken/fix/typo

Update README for french, spanish and italian language

mcp97 committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1237 from yjshi2015/main

Update trump.character.json - Enhance terminology in the project for clarity and inclusivity

mcp97 committed:
Merge branch 'develop' into tcm-twitter-character-template-issue

mcp97 committed:
Merge branch 'develop' into feat/express-payload-config

mcp97 committed:
Merge branch 'develop' into fix-cicd-package-visibility

mcp97 committed:
Merge pull request #1243 from luisalrp/fix-cicd-package-visibility

Fix visibility issue github image cicd

mcp97 committed:
Merge pull request #1242 from ai16z/tcm-twitter-character-template-issue

fix: twitterShouldRespondTemplate Fails When Defined as a String in JSON Character Config

mcp97 committed:
Merge branch 'develop' into feat/express-payload-config

mcp97 committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1202 from netdragonx/main

fix: optional chaining on search to avoid startup errors when search is not enabled

mcp97 committed:
Merge branch 'develop' into feat/express-payload-config

mcp97 committed:
Merge pull request #1245 from renlulu/feat/express-payload-config

feat: make express payload limit configurable

mcp97 committed:
Merge branch 'develop' into fix-ll

mcp97 committed:
Merge pull request #1250 from odilitime/fix-ll

fix: Fix local_llama key warning

lalalune committed:
Update trump.character.json

Remove a controversial line so we don't have to waste time on pointless old world political divisions

julienbrs committed:
doc: add Twitter automation label notice (#1253)

mcp97 committed:
Merge pull request #1254 from julienbrs/1253--add-twitter-automation-label-notice

mcp97 committed:
Merge pull request #1252 from ai16z/lalalune-patch-1

Commit:
mapping claude vertex

GeneFuture committed:
add tee-verifiable-log plugin api

gene-zhan committed:
Update package.json

shakkernerd committed:
Merge pull request #1258 from ai16z/tcm-claude-vertext-miss

fix: unsupported model provider: claude_vertex

odilitime committed:
Merge pull request #1246 from tomguluson92/tomguluson92-patch-2

feat: upgrade Tavily API with comprehensive input and constrain the token consumption

odilitime commented:
is this a duplicate of #1259 ?

derRizzMeister committed:
feat: add README_DE.md in docs directory

mcp97 committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1262 from derRizzMeister/main

feat: add README_DE.md in docs directory

HashWarlock commented on packages/plugin-tee-verifiable-log/README.md:
````diff
@@ -0,0 +1,33 @@
+```shell
+
+ # Optional, for simulator purposes if testing on mac or windows. Leave empty for Linux x86 machines.
+DSTACK_SIMULATOR_ENDPOINT="http://host.docker.internal:8090"
````
Change this to
```suggestion
TEE_MODE=DOCKER
```

HashWarlock replied:
@cpppppp7 can you add this small update, so I can add my approval?

HashWarlock commented on packages/client-direct/src/README.md:
````diff
@@ -43,3 +43,42 @@ curl -X GET "http://localhost:3000/fine-tune/8566c47a-ada8-441c-95bc-7bb07656c4c
 ```
+
+
+## Verifiable Attestations
````
Love this! 

I think the easiest way to insert into documentation is to add this as a section called **Enable Verifiable Logs** for the **Eliza in TEE** Doc `docs/docs/advanced/eliza-in-tee.md`. Then add a reference to your discord or discord user to contact about the implementation.

HashWarlock replied:
I think we can work on this together for formatting & further enhancements/tutorials down the road. Separate into to two Plugin Sections:
- TEE Plugin
   - Core Components
- TEE Verifiable Log Plugin
   - Core Components
- Tutorial
   - Enable Verifiable Log
- Conclusion
   - Mention contributors for implementation and who to reach out to learn more about Verifiable Logs in TEE

cpppppp7 replied:
No problem, let's work on this together. that’s exactly the idea behind building this plugin.

Let me explain further:  
- #1259 : This is a standalone log plugin that doesn’t modify other plugins or the client.  
- #1260 : This modifies the direct client to provide a query interface for verifiable logs.

@HashWarlock @odilitime

HashWarlock commented:
LGTM, outside of the 1 suggested edit. I can help document tomorrow and run a test to fix any bugs i may encounter

tobbelobb committed:
Ignore agent content

oxSaturn committed:
fix: fix ENABLE_ACTION_PROCESSING logic

shakkernerd committed:
fix: pnpm lockfile

shakkernerd committed:
Merge pull request #1273 from ai16z/fix/pnpm-lockfile

fix: pnpm lockfile

shakkernerd committed:
Revert "fix: pnpm lockfile"

shakkernerd committed:
Merge pull request #1275 from ai16z/revert-1273-fix/pnpm-lockfile

chore: Revert "fix: pnpm lockfile"

shakkernerd committed:
feat: init adapter base files

shakkernerd committed:
chore: install ioredis package

shakkernerd committed:
feat: redis client base class

shakkernerd committed:
feat: add redis url param

shakkernerd committed:
chore: add ioredis types

shakkernerd committed:
chore: pnpm lock file

shakkernerd committed:
fix: console -> elizaLogger

odilitime committed:
fix build phase when pnpm-lock is cleared

odilitime committed:
Merge pull request #1182 from SumeetChougule/main

fix: Fix client.push issue and update README for Slack client verification

odilitime committed:
Merge branch 'develop' into main

odilitime committed:
Merge pull request #1205 from tobbelobb/main

fix: write summary file before trying to cache it

odilitime committed:
Merge branch 'develop' into fix/ENABLE_ACTION_PROCESSING

odilitime committed:
Merge pull request #1268 from oxSaturn/fix/ENABLE_ACTION_PROCESSING

fix: fix ENABLE_ACTION_PROCESSING logic

shakkernerd committed:
chore: pnpm lock file

odilitime committed:
Merge branch 'develop' into fix-ts

shakkernerd committed:
Merge pull request #1283 from odilitime/fix-ts

fix: fix lockfile

shakkernerd committed:
Merge branch 'develop' into feat/redis_adapter

shakkernerd committed:
Merge branch 'develop' into feat/redis_adapter

odilitime committed:
remove unneeded package that PR1182 added

odilitime committed:
remove slack added by PR1182

shakkernerd committed:
Merge pull request #1289 from odilitime/fix-dev

chore: clean up merged PR1168

twilwa committed:
fix integrationTests.yaml to properly use repo env vars

shakkernerd committed:
Merge branch 'develop' into feat/redis_adapter

twilwa committed:
Merge branch 'develop' of https://github.com/ai16z/eliza into develop

* 'develop' of https://github.com/ai16z/eliza:
  remove slack added by PR1182
  remove unneeded package that PR1182 added
  fix build phase when pnpm-lock is cleared

shakkernerd committed:
Merge pull request #1279 from ai16z/feat/redis_adapter

feat: Redis Cache Implementation

shakkernerd committed:
Merge branch 'develop' into fix/integration-fix

shakkernerd committed:
Merge pull request #1291 from ai16z/fix/integration-fix

fix: integration tests fix

shakkernerd committed:
fix: pnpm lock file

shakkernerd committed:
Merge pull request #1292 from ai16z/fix/pnpm-lockfile

fix: pnpm lock file

shakkernerd committed:
fix: add missing claude vertex case to handleProvider

shakkernerd committed:
Merge pull request #1293 from ai16z/patch/handle_provider

fix: add missing claude vertex case to handleProvider

twilwa committed:
output checkable variable for conditional

shakkernerd committed:
feat: types for Cache Store

shakkernerd committed:
chore: add adapter-redis package

shakkernerd committed:
feat: add cache env to env example

shakkernerd committed:
chore: pnpm lock file

shakkernerd committed:
feat: add support for redis cache

shakkernerd committed:
Merge pull request #1294 from ai16z/fix/integration-tests

fix: output checkable variable for conditional

shakkernerd committed:
Merge branch 'develop' into feat/agent_redis_support

shakkernerd committed:
Merge pull request #1295 from ai16z/feat/agent_redis_support

feat: Add caching support for Redis

shakkernerd committed:
chore: bump version to 0.1.6-alpha.5

shakkernerd committed:
Merge pull request #1296 from ai16z/new_version_0.1.6-alpha5

chore: bump version to 0.1.6-alpha.5

shakkernerd committed:
Merge pull request #1290 from ai16z/develop

feat: Update main for v0.1.6-alpha.5

shakkernerd committed:
fix: remove clients from default character

shakkernerd committed:
Merge pull request #1297 from ai16z/fix/default_character

fix: remove clients from default character

shakkernerd committed:
fix: default character model to LLAMALOCAL

shakkernerd committed:
Merge pull request #1299 from ai16z/fix/default_character

fix: default character model to LLAMALOCAL

actions-user committed:
chore: update changelog

shakkernerd committed:
chore: Updated node version to 23.3.0 and pnpm version to 9.15.0

shakkernerd committed:
feat: release version 0.1.6

shakkernerd committed:
Merge pull request #1300 from ai16z/new_version_0.1.6

feat: release version 0.1.6

actions-user committed:
chore: update changelog

HashWarlock commented on packages/client-direct/src/index.ts:
```diff
@@ -69,6 +71,9 @@ export class DirectClient {
         const apiRouter = createApiRouter(this.agents, this);
         this.app.use(apiRouter);
 
+        const apiLogRouter = createVerifiableLogApiRouter(this.agents, this);
```
If `VERIFIABLE_LOGGING` is not enabled, will this fail to start the agent? or will it just fail the API call?

gene-zhan replied:
fail  the api call

gene-zhan committed:
Merge branch 'ai16z:main' into main

gene-sf committed:
merge main

gene-sf committed:
merge main

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback  main docs

gene-sf committed:
rollback main docs

HashWarlock commented:
@odilitime what are the @gene-sf changes? Is that some automation? I'm approving this PR now

gene-zhan commented:
> 



> ~@odilitime~ @gene-zhan what are the @gene-sf changes? Is that some automation? I'm approving this PR now
sorry，I have configured two GitHub accounts locally. The problem with @gene-sf is that I forgot to switch the configuration file when submitting the code.

I have completed the following operations:
- [x] According to your review, I have fixed the code.
- [x] Updated the readme that is easier to understand.
- [x] Merged the latest version of main code.

gene-zhan commented:
@shakkernerd  I noticed that PRs #1331  and #1333  were reopened and this PR closed, which has left me a bit confused about the current direction. Should I merge my changes into the develop branch, or is there a different branch we should treat as the standard? I’ve submitted #1369.  both [cpppppp7](https://github.com/cpppppp7) and I have been working diligently to make progress on this code contribution. I’d appreciate any clarification to ensure we align our efforts effectively. 😊

shakkernerd commented:
Hi @gene-zhan Thanks for bringing this to my attention.
PRs that were closed about 5hours ago were due to `develop` branch getting accidentally deleted.
And yes, the `develop` remains the standard to open PRs to.

Good to hear about your active contributions. 
Apologies for the confusion.

gene-zhan commented:
@shakkernerd  Thank you for your response! I accidentally submitted duplicate PRs (#1259 #1260  #1331, #1333, and #1369). I recommend prioritizing the merge of #1369, as this branch already includes the latest develop code. The other branches contain the same code but currently have conflicts that need to be resolved. Merging #1369 would help reduce additional workload. Thank you again for your understanding and support! 😊

gene-zhan commented:
@shakkernerd @odilitime @HashWarlock  How can I help facilitate the merging of this PR?

shakkernerd commented:
Hi @gene-zhan which PRs do you need attention on?

gene-zhan commented:
@shakkernerd  

I would like to request the following actions:

1. Close the following pull requests (PRs):
- #1259 
- #1260 
- #1331 
- #1333 
These PRs are no longer needed, and we can safely close them.

2. Merge PR #1369
Please proceed with merging PR #1369 into the develop branch as it contains the necessary updates/fixes we need.
Let me know if there are any issues or additional steps required. Thank you!

shakkernerd commented:
Okay, thanks for listing them out. I will get to it.

shakkernerd commented:
Closing in favor of #1369
